File: src/python/grpc/grpcio/cache.py
"""Redis connection management for the benchmark service."""

import os

import redis


_pool: redis.ConnectionPool | None = None


def get_client() -> redis.Redis:
    """Get or create the Redis client."""
    global _pool
    if _pool is None:
        redis_url = os.environ.get("REDIS_URL", "redis://localhost:6379/0")
        _pool = redis.ConnectionPool.from_url(
            redis_url,
            decode_responses=True,
            max_connections=10,
        )
    return redis.Redis(connection_pool=_pool)


def check_cache() -> str:
    """Check Redis connectivity and return status."""
    try:
        client = get_client()
        client.ping()
        return "connected"
    except Exception as e:
        return f"error: {e}"


def close_pool():
    """Close the Redis connection pool."""
    global _pool
    if _pool is not None:
        _pool.disconnect()
        _pool = None

File: src/dart/grpc/grpc-dart/lib/src/cache.dart
import 'dart:io';

import 'package:redis/redis.dart';

/// Redis connection manager for the benchmark service.
class Cache {
  Command? _command;
  RedisConnection? _connection;

  /// Redis URL from environment.
  String get _redisUrl =>
      Platform.environment['REDIS_URL'] ?? 'redis://localhost:6379';

  /// Get or create the Redis command.
  Future<Command> getCommand() async {
    if (_command == null) {
      _connection = RedisConnection();
      final uri = Uri.parse(_redisUrl);
      _command = await _connection!.connect(
        uri.host,
        uri.port,
      );
    }
    return _command!;
  }

  /// Check Redis connectivity and return status string.
  Future<String> checkCache() async {
    try {
      final cmd = await getCommand();
      await cmd.send_object(['PING']);
      return 'connected';
    } catch (e) {
      return 'error: $e';
    }
  }

  /// Close the Redis connection.
  Future<void> close() async {
    if (_connection != null) {
      _connection!.close();
    }
    _connection = null;
    _command = null;
  }
}
